src/query/api/v1/handler/prom/read.go

```diff
@@ -123,9 +124,20 @@ func (h *readHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	// NB (@shreyas): We put the FetchOptions in context so it can be
 	// retrieved in the queryable object as there is no other way to pass
 	// that through.
-	var resultMetadata block.ResultMetadata
+	//
+	// We also put a function into the context that allows callers to safely
+	// pass back result metadata concurrently so that they can be combined
+	// for later reporting.
+	var resultMetadataMutex sync.Mutex
+	resultMetadata := block.NewResultMetadata()
+	resultMetadataReceiveFn := func(m block.ResultMetadata) {
+		resultMetadataMutex.Lock()
+		defer resultMetadataMutex.Unlock()
+		resultMetadata = resultMetadata.CombineMetadata(m)
+
```

Comment:
nit: errant newline at the end and can probably skip the defer if we just unlock after the set